src/box/txn_limbo.c

```diff
@@ -40,6 +40,69 @@
 
 struct txn_limbo txn_limbo;
 
+static int
+synchro_request_write_without_error_handling(const struct synchro_request *req);
+
+static void
+txn_limbo_read_confirm(struct txn_limbo *limbo, int64_t lsn);
+
+/**
+ * Write a confirmation entry to the WAL. After it's written all the
+ * transactions waiting for confirmation may be finished.
+ */
+static int
+txn_limbo_write_confirm(struct txn_limbo *limbo)
+{
+	assert(!limbo->is_in_rollback);
+	struct synchro_request req = {
+		.type = IPROTO_RAFT_CONFIRM,
+		.replica_id = limbo->owner_id,
+		.lsn = limbo->volatile_confirmed_lsn,
+		.term = 0,
+		.confirmed_vclock = NULL,
+	};
+	return synchro_request_write_without_error_handling(&req);
+}
+
+static int
+txn_limbo_worker_bump_confirmed_lsn(struct txn_limbo *limbo)
+{
+	assert(limbo->is_in_rollback ||
+	       limbo->volatile_confirmed_lsn >= limbo->confirmed_lsn);
+	if (!txn_limbo_is_owned_by_current_instance(limbo) ||
+	    limbo->volatile_confirmed_lsn == limbo->confirmed_lsn)
+		return 0;
+	if (limbo->is_in_rollback)
+		return -1;
+	if (txn_limbo_write_confirm(limbo) != 0) {
+		diag_log();
+		return -1;
+	}
+
+	limbo->confirmed_lsn = limbo->volatile_confirmed_lsn;
+	vclock_follow(&limbo->confirmed_vclock, limbo->owner_id,
+		      limbo->confirmed_lsn);
+	txn_limbo_read_confirm(limbo, limbo->confirmed_lsn);
+	return 0;
+}
+
+static int
+txn_limbo_worker_f(va_list args)
+{
+	(void)args;
+	struct txn_limbo *limbo = fiber()->f_arg;
+	assert(limbo == &txn_limbo);
+	while (!fiber_is_cancelled()) {
+		fiber_check_gc();
+		ERROR_INJECT_YIELD(ERRINJ_TXN_LIMBO_WORKER_DELAY);
+		if (txn_limbo_worker_bump_confirmed_lsn(limbo) != 0)
+			fiber_sleep(0);
```

Comment:
Lets make the sleep non-zero. Otherwise in case of any long lasting WAL problems this will go into 100% CPU. Something like 1 second hardcoded would be enough. And in debug-build could make it 0.1 sec for faster tests.

Reply:
I was looking at this code:
https://github.com/tarantool/tarantool/blob/4a866f64d64c610a3c8441835fee3d8dda5eca71/src/box/raft.c#L128-L142

I guess you are right that we should back off in this case. But let's not have configuration-dependent code.